src/Microsoft.Azure.SignalR.AspNet/Infrastructure/SignalRMessageParser.cs

```diff
@@ -0,0 +1,216 @@
+                    case CommandType.Abort:
+                        yield break;
+                }
+            }
+
+            var segment = GetPayload(message);
+
+            // broadcast case
+            if (TryGetName(message.Key, PrefixHelper.HubPrefix, out var hubName))
+            {
+                yield return new HubMessage(hubName, new BroadcastDataMessage(excludedList: GetExcludedIds(message.Filter), payloads: GetPayloads(segment)), message);
+            }
+            // echo case
+            else if (TryGetName(message.Key, PrefixHelper.HubConnectionIdPrefix, out _))
+            {
+                // naming: hc-{HubName}.{ConnectionId}
+                // ConnectionId can NEVER contain .
+                var index = message.Key.LastIndexOf('.');
+                if (index < 0 || index == message.Key.Length - 1)
+                {
+                    throw new ArgumentException($"Key must contain '.' in between but it is not: {message.Key}");
+                }
+
+                var connectionId = message.Key.Substring(index + 1);
+
+                // Go through the app connection
+                yield return new AppMessage(new ConnectionDataMessage(connectionId, segment), message);
+            }
+            // group broadcast case
+            else if (TryGetName(message.Key, PrefixHelper.HubGroupPrefix, out _))
+            {
+                // naming: hg-{HubName}.{GroupName}, it as a whole is the group name per the JoinGroup implementation
+                // go through the app connection
+                // use groupName as the partitionkey so that commands towards the same group always goes into the same service connection
+                var groupName = message.Key;
+                yield return new AppMessage(new GroupBroadcastDataMessage(groupName, excludedList: GetExcludedIds(message.Filter), payloads: GetPayloads(segment)), message);
+            }
+            // user case
+            else if (TryGetName(message.Key, PrefixHelper.HubUserPrefix, out var userWithHubPrefix))
+            {
+                // naming: hu-{HubName}.{UserName}, HubName can contain '.' and UserName can contain '.'
+                // Go through all the possibilities
+                foreach (var (hub, user) in GetPossibleNames(userWithHubPrefix))
+                {
+                    // For old protocol, it is always single user per message https://github.com/SignalR/SignalR/blob/dev/src/Microsoft.AspNet.SignalR.Core/Infrastructure/Connection.cs#L162
+                    yield return new HubMessage(hub, new UserDataMessage(user, GetPayloads(segment)), message);
+                }
+            }
+            else
+            {
+                throw new NotSupportedException($"Message {message.Key} is not supported.");
+            }
+        }
+
+        private ReadOnlyMemory<byte> GetPayload(Message message)
+        {
+            IJsonWritable value = new PersistentResponse(m => false, tw => tw.Write("Cursor"))
```

Comment:
I think we should change "Cursor" to something more obscure so that nobody pays attention to it.

Reply:
Is it that anything is ok? how about empty string?

Reply:
Maybe null?